# Data Science Boot Camp

## Introduction to Python

<h2 style="color:maroon;"> Chapters </h2>
<ul>
    <li><a href="#intro" style="text-decoration:none;"><h3 style="color:black;">Introduction to Python</h3></a></li>
    <li><a href="#types" style="text-decoration:none;"><h3 style="color:black;">Types and Operations</h3></a></li>
    <li><a href="#syntax" style="text-decoration:none;"><h3 style="color:black;">Statements and Syntax</h3></a></li>
    <li><a href="#func" style="text-decoration:none;"><h3 style="color:black;">Functions and Files</h3></a></li>
    <li><a href="#oop" style="text-decoration:none;"><h3 style="color:black;">Object-Oriented Programming & Classes</h3></a></li>
</ul>



<div id="intro">
    <h2 style="color:maroon;"> Introduction to Python </h2>
    <ul>
        <li><a style="text-decoration:none;"><h3 style="color:black;">History of Python</h3></a></li>
        <li><a style="text-decoration:none;"><h3 style="color:black;">What is Python?</h3></a></li>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Installing & Running Python</h3></a></li>
    </ul>
</div>


<h3 style="color:maroon;">History of Python</h3>

* Invented in the Netherlands, early 90’s by Guido van Rossum<br>
<br>
* Named after “Monty Python”<br>
<br>
* Open sourced from the beginning<br>
<br>
* Considered a scripting language, but is much more<br>
<br>
* Scalable, object oriented and functional from the beginning<br>
<br>
* Used by Google from the beginning

* Monty Python (also collectively known as The Pythons) were a British surreal comedy group who created their sketch comedy show Monty Python's Flying Circus, which first aired on the BBC in 1969. Forty-five episodes were made over four series.


“Python is an experiment in how much freedom programmers need. Too much freedom and nobody can read another's code; too little and expressiveness is endangered.”
<br>
<p style="text-align:right;">- Guido van Rossum</p>

* Guido van Rossum ( born 31 January 1956) is a Dutch programmer best known as the author of the Python programming language, for which he is the "Benevolent Dictator For Life" (BDFL), which means he continues to oversee Python development, making decisions when necessary. From 2005 to December 2012, he worked at Google, where he spent half of his time developing the Python language. In January 2013, he started working for Dropbox.

<h3 style="color:maroon;">Python 1</h3>
* Python 1.0 - January 1994<br>
<br>
* Python 1.5 - December 31, 1997<br>
<br>
* Python 1.6 - September 5, 2000<br>

<h3 style="color:maroon;">Python 2</h3>
* Python 2.0 - October 16, 2000<br>
<br>
* Python 2.1 - April 17, 2001<br>
<br>
* Python 2.2 - December 21, 2001<br>
<br>
* Python 2.3 - July 29, 2003<br>
<br>
* Python 2.4 - November 30, 2004<br>
<br>
* Python 2.5 - September 19, 2006<br>
<br>
* Python 2.6 - October 1, 2008<br>
<br>
* Python 2.7 - July 3, 2010<br>

<h3 style="color:maroon;">Python 3</h3>
* Python 3.0 - December 3, 2008<br>
<br>
* Python 3.1 - June 27, 2009<br>
<br>
* Python 3.2 - February 20, 2011<br>
<br>
* Python 3.3 - September 29, 2012<br>
<br>
* Python 3.4 - March 16, 2014<br>
<br>
* Python 3.5 - September 13, 2015<br>
<br>
* Python 3.6 - December 23, 2016

<h3 style="color:maroon;">What is Python?</h3>

* Python is a open source computer programming language.<br>
<br>
* Python is an interpreted language.<br>
<br>
* Python is dynamically typed.<br>
<br>
* Python is well suited to object oriented programming.<br>
<br>
* Python is simple, easy to learn syntax emphasizes readability.<br>
<br>
* Python makes difficult things easy so programmers can focus on overriding algorithms and structures rather than low level details.

- Python is an interpreted language. That means that, unlike languages like C and its variants, Python does not need to be compiled before it is run. Other interpreted languages include PHP and Ruby.
- Python is dynamically typed, this means that you don't need to state the types of variables when you declare them or anything like that. You can do things like x=111 and then x="I'm a string" without error.
- Like JavaScript, Python supports a programming style that uses simple functions and variables without engaging in class definitions.

* The Zen of Python is a collection of 20 software principles that influences the design of Python Programming Language, only 19 of which were written down—around June 1999 by Tim Peters.

In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


Some of the platforms that Python supports:

* Amiga Research OS (AROS)<br>
<br>
* Application System 400 (AS/400)<br>
<br>
* BeOS<br>
<br>
* Linux<br>
<br>
* Mac OS X (comes pre-installed with the OS)<br>
<br>
* Microsoft Disk Operating System (MS-DOS)<br>
<br>
* Playstation<br>
<br>
* Solaris<br>
<br>
* Windows 32-bit / 64-bit<br>
<br>
* Windows CE/Pocket PC

Why Do People Use Python?

* Software quality<br>
<br>
* Developer productivity<br>
<br>
* Program portability<br>
<br>
* Support libraries<br>
<br>
* Component integration<br>
<br>
* Enjoyment

What Can I Do with Python?

* Systems Programming<br>
<br>
* GUIs<br>
<br>
* Internet Scripting<br>
<br>
* Component Integration<br>
<br>
* Database Programming<br>
<br>
* Rapid Prototyping

What Can I Do More?

* Data Analysis<br>
<br>
* Gaming<br>
<br>
* Image Processing<br>
<br>
* Artificial Intelligence<br>
<br>
* Data Visualization<br>
<br>
* Scientific Programming

`Systems Programming`<br>
* Python’s built-in interfaces to operating-system services make it ideal for writing portable, maintainable system-administration tools and utilities (sometimes called shell tools). Python programs can search files and directory trees, launch other programs, do parallel processing with processes and threads, and so on.
* Python’s standard library comes with POSIX bindings and support for all the usual OS tools: environment variables, files, sockets, pipes, processes, multiple threads, regular expression pattern matching, command-line arguments, standard stream interfaces, shell-command launchers, filename expansion, zip file utilities, XML and JSON parsers, CSV file handlers, and more. In addition, the bulk of Python’s system interfaces are designed to be portable; for example, a script that copies directory trees typically runs unchanged on all major Python platforms.<br>
`GUIs`<br>
* Python’s simplicity and rapid turnaround also make it a good match for graphical user interface programming on the desktop. Python comes with a standard object-oriented interface to the Tk GUI API called tkinter (Tkinter in 2.X) that allows Python programs to implement portable GUIs with a native look and feel. Python/tkinter GUIs run unchanged on Microsoft Windows, X Windows (on Unix and Linux), and the Mac OS (both Classic and OS X).<br>
`Internet Scripting`<br>
* Python comes with standard Internet modules that allow Python programs to perform a wide variety of networking tasks, in client and server modes. Scripts can communicate over sockets; extract form information sent to server-side CGI scripts; transfer files by FTP; parse and generate XML and JSON documents… In addition, full-blown web development framework packages for Python, such as Django, TurboGears, web2py, Pylons, Zope, and WebWare, support quick construction of full-featured and production-quality websites with Python.<br>
`Component Integration`<br>
* Python’s ability to be extended by and embedded in C and C++ systems makes it useful as a flexible glue language for scripting the behavior of other systems and components. For instance, integrating a C library into Python enables Python to test and launch the library’s components, and embedding Python in a product enables onsite customizations to be coded without having to recompile the entire product (or ship its source code at all).<br>
`Database Programming`<br>
* For traditional database demands, there are Python interfaces to all commonly used relational database systems—Sybase, Oracle, Informix, ODBC, MySQL, PostgreSQL, SQLite, and more. The Python world has also defined a portable database API for ac- cessing SQL database systems from Python scripts.<br>
`Rapid Prototyping`<br>
* Unlike some prototyping tools, Python doesn’t require a complete rewrite once the prototype has solidified. Parts of the system that don’t require the efficiency of a language such as C++ can remain coded in Python for ease of maintenance and use.<br>
`Data Analysis`<br>
* Powerful statistical and numerical packages. NumPy and pandas allow you to read/manipulate data efficiently and easily. Scikit-learn allows you to train and apply machine learning algorithms to your data and make predictions. 
* PyMySQL allows you to easily connect to MySQL database, execute queries and extract data.
* BeautifulSoup to easily read in XML and HTML type data which is quite common nowadays.
* Natural language analysis with the NLTK package.<br>
`Gaming`<br>
* Game programming and multimedia with pygame, cgkit, pyglet, PySoy, Panda3D, and others<br>
`Image Processing`<br>
* Image processing with PIL and its newer Pillow fork, PyOpenGL, Blender, Maya, and more<br>
`Artificial Intelligence`<br>
* Artificial intelligence with the PyBrain neural net library and the Milk machine learning toolkit
* Tensorflow for some neural network<br>
`Data Visualization`<br>
* Data visualization with Mayavi, matplotlib, VTK, VPython, and more<br>
`Scientific Programming`<br>
* Scientific Programming has grown to become one of Python’s most compelling use cases. Prominent here, the NumPy high-performance numeric programming extension for Python mentioned earlier in- cludes such advanced tools as an array object, interfaces to standard mathematical libraries, and much more.


<h3 style="color:maroon;">Installing Python</h3>

* Python is pre-installed on most Unix systems, including Linux and MAC OS X<br>
<br>
* The pre-installed version may not be the most recent one.<br>
<br>
* You can download from “http://python.org/download/”.<br>
<br>
* Also you can check out Anaconda.<br>
<br>
* Anaconda is a Python distribution that includes nearly 200 packages for data science. ( NumPy, SciPy, pandas, Jupyter, Matplotlib)

* There are several options for an IDE<br>
<br>
* The best Python IDEs for data science are “Spyder”, “PyCharm”, “Rodeo”, “Jupyter Notebook” and “Atom Text Editor”.<br>
<br>
* We will use Jupyter formerly known as Ipython Notebook in this boot camp.

“The best Python IDEs” paper from datacamp.com → https://www.datacamp.com/community/tutorials/data-science-python-ide

<h3 style="color:maroon;">Running Python</h3>

* For running interactively on Unix terminals or Windows Powershell type ‘python’.<br
<br>
> Python prompts with ‘>>>’.<br>
\>\>\> 2 + 2<br>
4<br>
    * To exit Python (not Idle):
    * In Unix, type CONTROL-D
    * In Windows, type CONTROL-Z + <Enter>
    * Evaluate exit() 

* For running python scripts via the python interpreter type ‘python’ and script name.
> python bootcamp.py 
* Make a python file directly executable by:
>Adding the appropriate path to your python interpreter as the first line of your file (#!/usr/bin/python)<br>
Making the file executable (chmod a+x bootcamp.py)<br>
Invoking file from Unix command line (bootcamp.py)<br>
Or you can use your IDE for running.<br>




<div id="types">
    <h2 style="color:maroon;"> Types and Operations </h2>
    <ul>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Variables</h3></a></li>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Data Types</h3></a></li>
    </ul>
</div>


<h3 style="color:maroon;">Variables</h3>

* Unlike other programming languages, Python does not need specific declaration of the data type to reserve memory space.<br>
<br>
* Values are assigned to variables with an equals sign ( “=” ).<br>

In [2]:
age = 3  # We assign the variable age a integer value
pi = 3.14  # We give the variable grade a floating point value
mood = "happy"  # We give the variable mood an string value


* Notice that each variable contains a different data type, however, we never had to explicitly declare each of their data types.

* You can also assign multiple variables a value at the same time.

In [3]:
a = b = c = 3

* You can also assign multiple variables on the same line with different values and even different data types:

In [4]:
age, pi, mood = 3, 3.14, 'happy'

* I'd recommend doing whatever you feel is the most simple to understand.

* The python language defines 2 types of variables:<br>
`Instance Variables` and `Local Variables`<br>

* An instance variable is visible to all functions, constructors and blocks, within the file that it is declared in.

In [5]:
a = 3
def testGlobalVariable():
    print(a)
testGlobalVariable()

3


* Local variables are only declared within functions, blocks or constructors. They are only created when the function, block or constructor is created, and then destroyed as soon as the function ends.

In [6]:
a = 3
def testLocalVariable():
    x = 5
testLocalVariable()
print(x)

NameError: name 'x' is not defined

* For having user input in python there is a function exist: raw_input

In [7]:
v = raw_input("Enter a value for me")

print("You entered:" + v)

type(v)

Enter a value for me5
You entered:5


str

<h3 style="color:maroon;">Data Types</h3>

* When you create a variable, the data stored within it can be of many different data types.<br>
<br>
* Python has multiple different data types that you can use to store different types of values, each of which will have its own set of unique operations that you can use to manipulate and work with the variables.<br>
<br>
* Python has 5 standard core Data Types:<br>
`Numbers`, `Strings`, `Lists`, `Tuples`, `Dictionary`

<h2 style="color:maroon;"> Data Types </h2>
<ul>
    <li><a style="text-decoration:none;"><h3 style="color:maroon;">Numbers</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Strings</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Lists</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Tuples</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Dictionary</h3></a></li>
</ul>




<h3 style="color:maroon;">Numbers</h3>

* Number data types store numbers.<br>
<br>
* They are created when you create a variable that holds a numerical value.<br>
<br>
* Python supports 4 types of number values:<br>
`int`, `long`, `float` and `complex`


* Int data type is a 32-bit signed two’s complement integer.<br>
<br>
* It’s maximum value is:<br>
`2,147,483,647 (2^31 -1)`<br>
<br>
* It’s minimum value is:<br>
`- 2,147,483,648 (-2^31)`<br>
<br>
* Int is the most common used data type

In [8]:
import sys
intSample = sys.maxint
print(intSample) #unsigned 2^63 - 1
print(type(intSample))
intSample = intSample + 1
print(type(intSample))

9223372036854775807
<type 'int'>
<type 'long'>


* Long data type is a 64-bit signed two’s complement integer.<br>
<br>
* It’s maximum value is:<br>
`9,223,372,036,854,775,807 (2^63 -1)`<br>
<br>
* It’s minimum value is:<br>
`-9,223,372,036,854,775,808 (-2^63)`<br>
<br>
* This data type is used when a larger number is required to work with than would be possible with an int

In [9]:
longSample = 12351235L  # has a value of 12351235
print(longSample)
print(type(longSample))

12351235
<type 'long'>


* The floating point data type is a double-precision 64-bit IEEE 754 floating point.<br>
<br>
* The min and max values are too large to discuss here.<br>
<br>
* For detailed information import sys and type sys.float_info<br>

In [10]:
floatSample = 3.14
print(floatSample)
print(type(floatSample))

3.14
<type 'float'>


In [11]:
import sys
sys.float_info


# v = sys.float_info.max

# if that's not big enough then there is another option :)
# infinity = float("inf")

# infinity > v



sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

* Complex numbers consist of ordered pairs of real floating-point numbers represented by x + yj<br>
<br>
* x is the real floating number, while yj is the imaginary part of it.<br>
<br>

In [12]:
complexSample = 3.14j
print(complexSample)
print(type(complexSample))

3.14j
<type 'complex'>


* Beginning with Python 2.0, the set of additional assignment statement formats became available. 
* Known as augmented assignments, and borrowed from the C language.

In [13]:
x = 10
y = 5
x = x + y   # Traditional form
x += y      # Newer augmented form

In [14]:
x += y
x *= y
x %= y
x &= y
x ^= y
x <<= y
x -= y
x /= y
x **= y
x |= y
x >>= y
x //= y

<table>
<thead>
<tr>
<th style="text-align:left;">Operator</th>
<th style="text-align:left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">lambda</td>
<td style="text-align:left;">Lambda expression</td>
</tr>
<tr>
<td style="text-align:left;">if – else</td>
<td style="text-align:left;">Conditional expression</td>
</tr>
<tr>
<td style="text-align:left;">or</td>
<td style="text-align:left;">Boolean OR</td>
</tr>
<tr>
<td style="text-align:left;">and</td>
<td style="text-align:left;">Boolean AND</td>
</tr>
<tr>
<td style="text-align:left;">not x</td>
<td style="text-align:left;">Boolean NOT</td>
</tr>
<tr>
<td style="text-align:left;">in, not in, is, is not, <, <=, >, >=, <>, !=, ==</td>
<td style="text-align:left;">Comparisons, including membership tests and identity tests</td>
</tr>
<tr>
<td style="text-align:left;">|</td>
<td style="text-align:left;">Bitwise OR</td>
</tr>
<tr>
<td style="text-align:left;">^</td>
<td style="text-align:left;">Bitwise XOR</td>
</tr>
<tr>
<td style="text-align:left;">&</td>
<td style="text-align:left;">Bitwise AND</td>
</tr>
<tr>
<td style="text-align:left;"><<, >></td>
<td style="text-align:left;">Shifts</td>
</tr>
<tr>
<td style="text-align:left;">+, -</td>
<td style="text-align:left;">Addition and subtraction</td>
</tr>
<tr>
<td style="text-align:left;">*, /, //, %</td>
<td style="text-align:left;">Multiplication, division, remainder</td>
</tr>
<tr>
<td style="text-align:left;">+x, -x, ~x</td>
<td style="text-align:left;">Positive, negative, bitwise NOT</td>
</tr>
<tr>
<td style="text-align:left;">**</td>
<td style="text-align:left;">Exponentiation</td>
</tr>
<tr>
<td style="text-align:left;">x[index], x[index:index], x(arguments...), x.attribute</td>
<td style="text-align:left;">Subscription, slicing, call, attribute reference</td>
</tr>
<tr>
<td style="text-align:left;">(expressions...), [expressions...], {key: value...}, `expressions...`</td>
<td style="text-align:left;">Binding or tuple display, list display, dictionary display, string conversion</td>
</tr>
</tbody>
</table>

* Python provides a built-in functions. We can examine this functions in 3 category for numbers:<br>
`Math Functions`<br>
<br>
`Conversion Functions`<br>
<br>
`Collection Functions`<br>

<h3 style="color:maroon;">Numbers (Math Functions)</h3>

* The formal definitions of mathematical built-in functions are provided below.

* abs ( number ) → number<br>
Return the absolute value of the argument, | x |.

In [15]:
abs(-10)

10

* pow ( x , y ) → number<br>
Raise x to the y power.

In [16]:
print(pow(2,3))
print(pow(2,3,5))

8
3


* round ( number , [ ndigits ] ) → float<br>
Round number to ndigits beyond the decimal point.

In [17]:
round(3.141592, 2)

3.14

* cmp ( x , y ) → integer<br>
Compare x and y , returning a number. 

In [18]:
print(cmp(10,5))
print(cmp(5,5))
print(cmp(5,10))

1
0
-1


<h3 style="color:maroon;">Numbers (Conversion Functions)</h3>

* The conversion functions provide alternate representations for numeric values.

* hex ( number ) → string<br>
Create a hexadecimal string representation of number . A leading '0x' is placed on the string as a reminder that this is hexadecimal.


In [19]:
hex(24)

'0x18'

* oct ( number ) → string<br>
Create a octal string representation of number . A leading '0' is placed on the string as a reminder that this is octal not decimal. 

In [20]:
oct(24)

'030'

* int ( string , [ base ]) → integer<br>
Generates an integer from the string x . If base is supplied, x must be in the given base. If base is omitted, x must be decimal.

In [21]:
print(int("100"))
print(type(int("100")))

100
<type 'int'>


In [22]:
int("010111",2)

23

* str ( object ) → string<br>
Generate a string representation of the given object.

In [23]:
print(str(200))
print(type(str(200)))

200
<type 'str'>


<h3 style="color:maroon;">Numbers (Collection Functions)</h3>

* These are two built-in functions which operate on a collection of data elements.

In [24]:
listSample = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

* max ( sequence ) → value<br>
Return the largest value in sequence

In [25]:
max(listSample)

9

* min ( sequence ) → value<br>
Return the smallest value in sequence

In [26]:
min(listSample)

0

* Python’s standard library is very extensive, offering a wide range of modules. Let’s look at 2 of them:<br>
<br>
`math Module`<br>
<br>
`random Module`

<h3 style="color:maroon;">Numbers (Math Module)</h3>

* Math Module provides access to the mathematical functions defined by the C standard.<br>
<br>
* These functions cannot be used with complex numbers; use the functions of the same name from the “cmath” module if you require support for complex numbers.<br>
<br>
* All return values are floats.

* Let’s check out some of the functions that provided by this module:

In [27]:
import math

* math.ceil (x) → float<br>
Return the ceiling of x as a float, the smallest integer value greater than or equal to x.

In [28]:
math.ceil(1.3)

2.0

* math.factorial(x) → float<br>
Return x factorial. Raises ValueError if x is not integral or is negative.

In [29]:
math.factorial(5)

120

* math.floor(x) → float<br>
Return the floor of x as a float, the largest integer value less than or equal to x.

In [30]:
math.floor(1.9)

1.0

* math.isnan(x) → boolean<br>
Check if the float x is a NaN (not a number)

In [31]:
math.isnan(0.9)

False

* math.pi<br>
The mathematical constant π = 3.141592…, to available precision.

In [32]:
math.pi

3.141592653589793

* math.e<br>
The mathematical constant e = 2.718281…, to available precision.

In [33]:
math.e

2.718281828459045

* math.exp(x) → float<br>
Return e**x

In [34]:
math.exp(3)

20.085536923187668

* math.log(x [, base]) → float<br>
With one argument, return the natural logarithm of x (to base e).
With two arguments, return the logarithm of x to the given base, calculated as log(x)/log(base).

In [35]:
print(math.log(20.085536923187668))
print(math.log(64,2))

3.0
6.0


* math.log10(x) → float<br>
Return the base-10 logarithm of x. This is usually more accurate than log(x, 10).

In [36]:
math.log10(100)

2.0

* math.pow(x, y) → float<br>
Return x raised to the power y. Unlike the built-in ** operator, math.pow() converts both its arguments to type float.

In [37]:
math.pow(2,6)

64.0

* math.sin(x) → float<br>
Return the sine of x radians.

In [38]:
math.sin(15)

0.6502878401571169

<h3 style="color:maroon;">Numbers (Random Module)</h3>

* random.random()<br>
Return the next random floating point number in the range [0.0, 1.0).

In [39]:
import random
random.random()

0.07508753463423534

* random.uniform(a, b) <br>
Return a random floating point number N such that a <= N <= b

In [40]:
random.uniform(1,10) 

6.925986237278145

* random.randint(a, b)<br>
Return a random integer N such that a <= N <= b.

In [41]:
random.randint(1, 10)

2

* random.randrange(start, stop[, step])<br>
Return a randomly selected element from range(start, stop, step).

In [42]:
random.randrange(0, 101, 2) 

12

* random.choice(seq)<br>
Return a random element from the non-empty sequence seq

In [43]:
random.choice('abcdefghij') 

'j'

* random.shuffle(x)<br>
Shuffle the sequence x in place. 

In [44]:
items = [1, 2, 3, 4, 5, 6, 7]
random.shuffle(items)
items

[6, 5, 1, 4, 7, 2, 3]

* random.sample(population, k)<br>
Return a k length list of unique elements chosen from the population sequence.<br>
Used for random sampling without replacement.

In [45]:
random.sample([1, 2, 3, 4, 5],  3)

[5, 3, 1]

<h2 style="color:maroon;"> Data Types </h2>
<ul>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Numbers</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:maroon;">Strings</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Lists</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Tuples</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Dictionary</h3></a></li>
</ul>




<h3 style="color:maroon;">Strings</h3>

* A string is identified as a continuous set of characters within a set of quotation marks.<br>
<br>
* Python allows strings to be declared with in single ('  ') and double (" ") quotes.<br>
<br>
* Also you can think strings as a immutable sequances in python.<br>

In [46]:
hello = "Hello World"
print(len(hello)) # return the length of string
character = hello[0] # return first character of string
print(character)

11
H


In [47]:
hello[1] = "o" # Immutable objects cannot be changed

TypeError: 'str' object does not support item assignment

* Also you can splice strings:

In [48]:
print(hello[0:2])
print(hello[:5])
print(hello[6:])
print(hello[:-2])
print(hello[0:-1:2])

He
Hello
World
Hello Wor
HloWr


* Strings support concatenation with the plus sign (joining two strings into a new string)

In [49]:
hello = "Hello"
world = "World"
msg = hello + " " + world
msg

'Hello World'

* Strings also support repetition

In [50]:
hello * 2

'HelloHello'

* You can expand strings to a list

In [51]:
listSample = list(msg)
listSample

['H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd']

* And you can join list elements to string

In [52]:
newMsg = ''.join(listSample)
newMsg

'Hello World'

In [53]:
newMsgWithDots = '.'.join(listSample)
newMsgWithDots

'H.e.l.l.o. .W.o.r.l.d'

* Operation we’ve looked so far is really a sequence operation. These operations will work on other sequences in Python. <br>
Now let’s check strings specific methods

In [54]:
hello = "Hello"
print(hello.find("lo")) # return index of substring 
hi = hello.replace("ello", "i")
print(hi)

3
Hi


* Split a string into substrings on a delimiter (handy as a simple form of parsing)<br>

In [55]:
line = "cat,dog,bird"
arr =  line.split(',')
arr

['cat', 'dog', 'bird']

* Upper and lowercase conversions:

In [56]:
print(hello.upper())
print(hello.lower())

HELLO
hello


* Content tests:

In [57]:
textSample = "world"
digitSample = "01010"
print(textSample.isalpha())
print(digitSample.isdigit())

True
True


* You can import re for pattern matching:

In [58]:
import re
pattern = 'Hello[ \t]*(.*)world'
match = re.match(pattern, 'Hello Python world')
match.group(1)

'Python '

* Strings also support an advanced substitution operation known as formatting:

In [59]:
'%s, dogs, and %s' % ('cats', 'birds') # Works on all versions

'cats, dogs, and birds'

In [60]:
'{0}, dogs, and {1}'.format('cats', 'birds') # Works on 2.6+, 3.0+

'cats, dogs, and birds'

In [61]:
'{}, dogs, and {}'.format('cats', 'birds') # Works on 2.7+, 3.1+

'cats, dogs, and birds'

<h2 style="color:maroon;"> Data Types </h2>
<ul>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Numbers</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Strings</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:maroon;">Lists</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Tuples</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Dictionary</h3></a></li>
</ul>




<h3 style="color:maroon;">Lists</h3>

* List object is the most general sequence provided by Python.<br>
<br>
* Lists have no fixed size.<br>
<br>
* Lists are mutable, can be modified in place<br>

* You can have a list hold multiple different types of data

In [62]:
listSample = [3,'happy',3.14]

* They are sequences, lists support all the sequence operations we looked at strings.

In [63]:
len(listSample)

3

* We can index, slice, and so on, just as for strings

In [64]:
print(listSample[0])
print(listSample[:-1]) # return a new list
print(listSample + [5,6]) # return a new list
print(listSample * 2) # return a new list

3
[3, 'happy']
[3, 'happy', 3.14, 5, 6]
[3, 'happy', 3.14, 3, 'happy', 3.14]


* Lists can grow and shrink on demand

In [65]:
listSample.append('cats') # add item to last
listSample

[3, 'happy', 3.14, 'cats']

In [66]:
listSample.pop() # pick & delete last item
listSample

[3, 'happy', 3.14]

In [67]:
listSample.pop(0) #  pick & delete item on index 0
listSample

['happy', 3.14]

In [68]:
listSample.insert(1, "dogs") # insert at given index
listSample

['happy', 'dogs', 3.14]

In [69]:
listSample.remove('dogs') # remove the first ‘dogs’ appearance in the list
listSample

['happy', 3.14]

* Most list methods change the list object in place, instead of creating a new one:

In [70]:
M = ['bb', 'aa', 'cc']
M.sort()
print(M)
M.reverse()
print(M)

['aa', 'bb', 'cc']
['cc', 'bb', 'aa']


* The list methods make it very easy to use a list as a stack

In [71]:
stack = [3, 4, 5]
stack.append(6)
print(stack)
stack.append(7)
print(stack)
stack.pop() # 7
print(stack)
stack.pop() # 6
print(stack)

[3, 4, 5, 6]
[3, 4, 5, 6, 7]
[3, 4, 5, 6]
[3, 4, 5]


* It is also possible to use a list as a queue (First In, First Out)

In [72]:
queue = ["cats", "dogs", "birds"]
queue.append("cows")
print(queue)
queue.pop(0) # “cats
print(queue)
queue.pop(0) # “dogs”
print(queue)

['cats', 'dogs', 'birds', 'cows']
['dogs', 'birds', 'cows']
['birds', 'cows']


* It is also possible to concat lists with zip() function:<br>
The zip() function returns an iterator of tuples based on the iterable object.<br>
If no parameters are passed, zip() returns an empty iterator<br>
If a single iterable is passed, zip() returns an iterator of 1-tuples.


In [73]:
list1 = [1, 2, 3]
list2 = ['a', 'b', 'c', 'd']
result = zip()
print(result)
result = zip(list1)
print(result)
result = zip(list1, list2)
print(result)

[]
[(1,), (2,), (3,)]
[(1, 'a'), (2, 'b'), (3, 'c')]


In [74]:
newList1, newList2 = zip(*result)
print(newList1)
print(newList2)

(1, 2, 3)
('a', 'b', 'c')


* One nice feature of Python’s core data types is that they support arbitrary nesting:

In [75]:
M = [[1, 2, 3], 
     [4, 5, 6],
     [7, 8, 9]]
print(M)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


In [76]:
M[1]

[4, 5, 6]

In [77]:
M[1][2]

6

<h2 style="color:maroon;"> Data Types </h2>
<ul>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Numbers</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Strings</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Lists</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:maroon;">Tuples</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Dictionary</h3></a></li>
</ul>




<h3 style="color:maroon;">Tuples</h3>

* The tuple object is roughly like a list that cannot be changed.<br>
<br>
* Tuples are sequences, like lists, but they are immutable, like strings.<br>
<br>
* They're used to represent fixed collections of items.


* They are normally coded in parentheses instead of square brackets.

In [78]:
tupleSample = (1, 2, 3, 4)
print(tupleSample)
print(len(tupleSample))
print(tupleSample[0])
print(tupleSample + (5,6))

(1, 2, 3, 4)
4
1
(1, 2, 3, 4, 5, 6)


* Tuples also have type-specific callable methods, but not nearly as many as lists

In [79]:
print(tupleSample)
print(tupleSample.index(4))
print(tupleSample.count(4))

(1, 2, 3, 4)
3
1


* The primary distinction for tuples is that they cannot be changed once created.

In [80]:
tupleSample[2] = 4

TypeError: 'tuple' object does not support item assignment

In [81]:
tupleSample.append(7)

AttributeError: 'tuple' object has no attribute 'append'

<h2 style="color:maroon;"> Data Types </h2>
<ul>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Numbers</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Strings</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Lists</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:black;">Tuples</h3></a></li>
    <li><a style="text-decoration:none;"><h3 style="color:maroon;">Dictionary</h3></a></li>
</ul>




<h3 style="color:maroon;">Dictionary</h3>

* Dictionaries consist of key-value pairs.<br>
<br>
* A Dictionary consists of a name, followed by a key, which then hold a certain amount of values.<br>
<br>
* A dictionary works in a similar way to a real dictionary. You can sort of think of a key as if it were a word, and then the value as the definition of the word, or the value that the word holds.<br>
<br>
* A key can be almost any python data type<br>

In [82]:
dict1 = {} # declares an empty dictionary
dict1["first"] = 3
dict1["second"] = "happy"
dict1["pi"] = 3.14
dict1

{'first': 3, 'pi': 3.14, 'second': 'happy'}

In [83]:
dict2 = {"first": 3, "second": "happy", "pi": 3.14}
dict2

{'first': 3, 'pi': 3.14, 'second': 'happy'}

In [84]:
dict3 = dict(zip(['first', 'second', 'pi'], [3, 'happy', 3.14]))
dict3

{'first': 3, 'pi': 3.14, 'second': 'happy'}

In [85]:
print(dict1["first"]) # prints the value of key "first"
print(dict2.keys()) # prints all keys
print(dict3.values()) # prints all values

3
['second', 'pi', 'first']
['happy', 3.14, 3]


<div id="syntax">
    <h2 style="color:maroon;"> Statements and Syntax </h2>
    <ul>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Control Flow Statements</h3></a></li>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Loops</h3></a></li>
    </ul>
</div>



<h3 style="color:maroon;">Control Flow Statements</h3>

* The Python if statement is typical of if statements in most procedural languages.<br>
<br>
* It takes the form of an if test, followed by one or more optional elif (“else if”) tests and a final optional else block.<br>
<br>
* The tests and the else part each have an associated block of nested statements, indented under a header line

In [86]:
if 1:
    print('True')

True


* As you can see, whitespace is meaningful in Python<br>
<br>
* Use a newline to end a line of code.<br>
<br>
* No braces { } to mark blocks of code, use consistent indentation instead.<br>
<br>
* First line with less indentation is outside of the block,<br>
* First line with more indentation starts a nested block<br>


In [87]:
x = 10
if x > 0:
    print("positive")
elif x == 0:
    print("0")
else:
    print(negative)


positive


* There is no switch or case statement in Python.<br>
<br>
* But you can use a dictionary-based 'switch':<br>

In [88]:
choices = {'ham': 1.99, 'eggs': 0.99, 'milk': 1.10}
customerChoice = 'ham'
print(choices[customerChoice])

1.99


* Also you can use get method of dictionary:

In [89]:
choices = {'ham': 1.99, 'eggs': 0.99, 'milk': 1.10}
print(choices.get('milk', 'Bad choice'))
print(choices.get('butter', 'Bad choice'))

1.1
Bad choice


* and (&)<br>
If both the operands are true then condition becomes true.<br>
(a and b) is true.

In [90]:
if 1 & 0:
    print('True')
else:
    print('False')

False


* or (|)<br>
If any of the two operands are non-zero then condition becomes true.<br>
(a or b) is true.

In [91]:
if 1 or 0:
    print('True')
else:
    print('False')

True


* not (!)<br>
Used to reverse the logical state of its operand.<br>
Not(a and b) is false.

In [92]:
if not 0:
    print('True')
else:
    print('False')

True


<h3 style="color:maroon;">Loops</h3>

* Python has two two main looping constructs; `while` and `for`.

In [93]:
i=0
while i<10: # Loop test
    print(i) #Loop body
    i+=1
    
print("while loop finished")

0
1
2
3
4
5
6
7
8
9
while loop finished


* The for loop is a generic iterator in Python: it can step through the items in any ordered sequence or other iterable object

In [94]:
print(listSample)
for elements in listSample:
    print(elements) #Loop body

print("\n\nfor loop finished")

['happy', 3.14]
happy
3.14


for loop finished


* Also you can use range function:

In [95]:
print(range(10))
print(range(5,10))
print(range(10,0, -1))
print(range(0,10, 3))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[5, 6, 7, 8, 9]
[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
[0, 3, 6, 9]


In [96]:
sum = 0
for i in range(0,10):
    sum = sum + i
print(sum)

45


* Any sequence works in a for, as it’s a generic tool:

In [97]:
textList = ["Hello", "World"]
username = "python"
for i in textList: 
    print(i),
for i in username:
    print(i + ''), # Print statement ends with a comma means print inline

Hello World p y t h o n


* `Break`, `continue`, `pass`, and the Loop `else` are the simple statements you can use with loops

* The break statement causes an immediate exit from a loop.

In [98]:
sum = 0
while True:
    num = int(raw_input("enter a integer (0 for exit)"))
    if num == 0: break
    sum = sum + num
print(sum)


enter a integer (0 for exit)0
0


<div id="func">
    <h2 style="color:maroon;"> Functions and Files </h2>
    <ul>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Function Basics</h3></a></li>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Lambda Functions</h3></a></li>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Files</h3></a></li>
    </ul>
</div>

<h3 style="color:maroon;">Function Basics</h3>

* In simple terms, a function is a device that groups a set of statements so they can be run more than once in a program, a packaged procedure invoked by name.<br>
<br>
* Functions are the alternative to programming by cutting and pasting rather than having multiple redundant copies of an operation’s code

In [99]:
def times(x, y):  # Create and assign function
    return x * y  # Body executed when called
prod = times(2,8) # Call function
prod

16

* Variables that are defined inside a function body have a local scope, and those defined outside have a global scope.<br>
* This means that local variables can be accessed only inside the function in which they are declared, whereas global variables can be accessed throughout the program body by all functions.


In [100]:
z = 0 # This is global variable
def sum(x, y):
    z = x + y # This is local variable
    print("Inside the function local sum: " + str(z))
    return z 
sum(2, 2)
print("Outside the function global sum: " + str(z))

Inside the function local sum: 4
Outside the function global sum: 0


* For changing global variable in function:

In [101]:
z = 0
def sum(x, y):
    global z
    z = x + y
    return z 
sum(2, 2)
print(z)

4


* You can call a function by using the following types of formal arguments:<br>
<br>
`Required arguments`<br>
<br>
`Default arguments`<br>
<br>
`Keyword arguments`<br>

<h3 style="color:maroon;">Required Arguments</h3>

* Required arguments are the arguments passed to a function in correct positional order.

In [102]:
def sayHello(msg):
    print(msg)

sayHello() # call function with 0 arguments

TypeError: sayHello() takes exactly 1 argument (0 given)

<h3 style="color:maroon;">Default Arguments</h3>

* A default argument is an argument that assumes a default value if a value is not provided in the function call for that argument.<br>
<br>
* For some functions, you may want to make some parameters optional, you can use default arguments.


In [103]:
def sayHello(msg, times=1):
    print(msg * times)
    
sayHello("merhaba")
sayHello("Hello", 2)

merhaba
HelloHello


<h3 style="color:maroon;">Keyword Arguments</h3>

* When you use keyword arguments in a function call, the caller identifies the arguments by the parameter name.<br>
<br>
* If you have some functions with many parameters and you want to specify only some of them, you can use keyword arguments.


In [104]:
def func(x, y, z=1):
    print("x: " + str(x))
    print("y: " + str(y)) 
    print("z: " + str(z))
    print("------")
func(10, 20, 30)
func(y=10, x=3)

x: 10
y: 20
z: 30
------
x: 3
y: 10
z: 1
------


<h3 style="color:maroon;">Extra: Variable-length arguments</h3>

* You may need to process a function for more arguments than you specified while defining the function.<br>
<br>
* These arguments are called variable-length arguments and are not named in the function definition, unlike required and default arguments.

In [105]:
def total(a=5, *numbers, **animals):
    print('a = ' + str(a))
    for item in numbers:
        print('item: ' + str(item))
        
    for firstPart, secondPart in animals.items():
        print(firstPart, secondPart)

total(10,1,2,3,cats=5,dogs=10)

a = 10
item: 1
item: 2
item: 3
('cats', 5)
('dogs', 10)


<h3 style="color:maroon;">Lambda Functions</h3>

* Python supports the creation of anonymous functions at runtime, using a construct called "lambda".<br>
<br>
* This piece of code in the next cell shows the difference between a normal function definition ("f") and a lambda function ("g"):

In [106]:
def f (x): 
    return x**2
x = f(5)
print(x)

25


In [107]:
g = lambda x: x**2
x = g(5) 
print(x)

25


* Note that the lambda definition does not include a "return" statement it always contains an expression which is returned.

* In the following slide, we will first define a simple list of integer values, then use the standard functions filter(), map() and reduce() to do various things with that list. All of the three functions expect two arguments: A function and a list.

In [108]:
numbers = [0, 6, 9, 12, 14]

In [109]:
print filter(lambda x: x % 3 == 0, numbers)

[0, 6, 9, 12]


* filter() calls our lambda function for each element of the list, and returns a new list that contains only those elements for which the function returned "True".<br>
<br>
* In this case, we get a list of all elements that are multiples of 3. The expression x % 3 == 0 computes the remainder of x divided by 3 and compares the result with 0 

In [110]:
print map(lambda x: x * 2 + 10, numbers)

[10, 22, 28, 34, 38]


* map() is used to convert our list. The given function is called for every element in the original list, and a new list is created which contains the return values from our lambda function.<br>
<br>
* In this case, it computes 2 * x + 10 for every element.

In [111]:
print reduce(lambda x, y: x + y, numbers)

41


* reduce() is somewhat special. The "worker function" for this one must accept two arguments (we've called them x and y here), not just one. <br>
<br>
* The function is called with the first two elements from the list, then with the result of that call and the third element, and so on, until all of the list elements have been handled. This means that our function is called n-1 times if the list contains n elements.<br>
<br>
* The return value of the last call is the result of the reduce() construct.<br>
<br>
* In the above example, it simply adds the arguments, so we get the sum of all elements.

<h3 style="color:maroon;">Files</h3>

* When you’re working with Python, you don’t need to import a library in order to read and write files. It’s handled natively in the language.<br>
<br>
* The first thing you’ll need to do is use Python’s built-in open function to get a file object.<br>
<br>
* In short, the built-in open function creates a Python file object, which serves as a link to a file residing on your machine.

In [112]:
myFile = open('myfile.txt', 'w')
myFile.write("['cats', 'dogs', 'birds', 'cows']\n['dogs', 'birds', 'cows']\n['birds', 'cows']")

In [113]:
myFile = open('myfile.txt')
print(myFile)

<open file 'myfile.txt', mode 'r' at 0x103db1c90>


In [114]:
myFile = open('myfile.txt')
print(myFile.readline()) # Return first line of file

['cats', 'dogs', 'birds', 'cows']



In [115]:
myFile = open('myfile.txt')
print(myFile.read()) # Read all at once into string

['cats', 'dogs', 'birds', 'cows']
['dogs', 'birds', 'cows']
['birds', 'cows']


In [116]:
myFile = open('myfile.txt')
for line in myFile:
    print(line)

['cats', 'dogs', 'birds', 'cows']

['dogs', 'birds', 'cows']

['birds', 'cows']


In [117]:
myFile = open('myfile.txt', 'a')
myFile.write("\n['cows']")
myFile.close()

In [118]:
myFile = open('myfile.txt')
print(myFile.read())

['cats', 'dogs', 'birds', 'cows']
['dogs', 'birds', 'cows']
['birds', 'cows']
['cows']


* You can see open modes and descriptions in the next slide.

<table>
<thead>
<tr>
<th style="text-align:left;">Mode</th>
<th style="text-align:left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">r</td>
<td style="text-align:left;">Opens a file for reading only. The file pointer is placed at the beginning of the file. This is the default mode.</td>
</tr>
<tr>
<td style="text-align:left;">rb</td>
<td style="text-align:left;">Opens a file for reading only in binary format. The file pointer is placed at the beginning of the file. This is the default mode.</td>
</tr>
<tr>
<td style="text-align:left;">r+</td>
<td style="text-align:left;">Opens a file for both reading and writing. The file pointer placed at the beginning of the file.</td>
</tr>
<tr>
<td style="text-align:left;">rb+</td>
<td style="text-align:left;">Opens a file for both reading and writing in binary format. The file pointer placed at the beginning of the file.</td>
</tr>
<tr>
<td style="text-align:left;">w</td>
<td style="text-align:left;">Opens a file for writing only. Overwrites the file if the file exists. If the file does not exist, creates a new file for writing.</td>
</tr>
<tr>
<td style="text-align:left;">wb</td>
<td style="text-align:left;">Opens a file for writing only in binary format. Overwrites the file if the file exists. If the file does not exist, creates a new file for writing.</td>
</tr>
<tr>
<td style="text-align:left;">w+</td>
<td style="text-align:left;">Opens a file for both writing and reading. Overwrites the existing file if the file exists. If the file does not exist, creates a new file for reading and writing.</td>
</tr>
<tr>
<td style="text-align:left;">wb+</td>
<td style="text-align:left;">Opens a file for both writing and reading in binary format. Overwrites the existing file if the file exists. If the file does not exist, creates a new file for reading and writing.</td>
</tr>
<tr>
<td style="text-align:left;">a</td>
<td style="text-align:left;">Opens a file for appending. The file pointer is at the end of the file if the file exists. That is, the file is in the append mode. If the file does not exist, it creates a new file for writing.</td>
</tr>
<tr>
<td style="text-align:left;">ab</td>
<td style="text-align:left;">Opens a file for appending in binary format. The file pointer is at the end of the file if the file exists. That is, the file is in the append mode. If the file does not exist, it creates a new file for writing.</td>
</tr>
<tr>
<td style="text-align:left;">a+</td>
<td style="text-align:left;">Opens a file for both appending and reading. The file pointer is at the end of the file if the file exists. The file opens in the append mode. If the file does not exist, it creates a new file for reading and writing.</td>
</tr>
<tr>
<td style="text-align:left;">ab+</td>
<td style="text-align:left;">Opens a file for both appending and reading in binary format. The file pointer is at the end of the file if the file exists. The file opens in the append mode. If the file does not exist, it creates a new file for reading and writing.
</td>
</tr>
</tbody>
</table>

* Example Write in CSV:

In [119]:
import csv
with open('sample.csv', 'wb') as csvfile:
    writer = csv.writer(csvfile, delimiter=',')
    writer.writerows([['cats', 'dogs', 'birds']])
    writer.writerows([['cows', 'birds', 'cats']])

* Example Read in CSV:

In [120]:
import csv
with open('sample.csv', 'rb') as csvfile:
    reader = csv.reader(csvfile, delimiter=',')
    for row in reader:
        print(row)

['cats', 'dogs', 'birds']
['cows', 'birds', 'cats']


* Example Write in JSON:

In [121]:
import json
weird_json = {"3": "Python", "2": "World", "1": "Hello"}
with open('sample.json', 'w') as jsonfile:
    json.dump(weird_json, jsonfile)

* Example Read in JSON:

In [122]:
import json
with open('sample.json') as jsonfile:
    data = json.load(jsonfile)
    print(data)

{u'1': u'Hello', u'3': u'Python', u'2': u'World'}


* When load json you can see ‘u’ prefix. The u prefix means that those strings are unicode rather than 8-bit strings. When you really use the string, it won't appear in your data.  

In [123]:
print(data['1'])
print(type(data['1']))
print(type(data['1'].encode('utf-8')))

Hello
<type 'unicode'>
<type 'str'>


<div id="oop">
    <h2 style="color:maroon;">Object-Oriented Programming & Classes</h2>
    <ul>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Object Oriented Programming (OOP)</h3></a></li>
        <li><a style="text-decoration:none;"><h3 style="color:black;">Classes</h3></a></li>
    </ul>
</div>


<h3 style="color:maroon;">Object Oriented Programming (OOP)</h3>

* In all the examples we look at it till now, is around functions i.e. blocks of statements which manipulate data. This is called the procedure-oriented way of programming.<br>
<br>
* There is another way of organizing your program which is to combine data and functionality and wrap it inside something called an object. This is called the object oriented programming paradigm.<br>
<br>
* Most of the time you can use procedural programming, but when writing large programs or have a problem that is better suited to this method, you can use object oriented programming techniques.

<h3 style="color:maroon;">Classes</h3>

* Classes and objects are the two main aspects of object oriented programming.<br>
<br>
* A class creates a new type where objects are instances of the class.<br>
<br>
* Objects are a representation of the real world objects like cars, dogs, peoples, etc.<br>
<br>
* The objects share two main characteristics: data and behavior.

* For Example, Cars have data like number of wheels, number of doors, seating capacity and also have behavior: accelerate, stop, show how much fuel is missing and so many other.

* Data = Attributes
* Behavior = Methods

* A class is a blueprint, a model for its objects, a way to define attributes and behavior.

Let’s look Python syntax for classes:

In [124]:
class Vehicle:
    pass
car = Vehicle() # Creating an instance

* Class methods have only one specific difference from ordinary functions:
> * Class methods have a extra parameter, but you do not give a value for this parameter when you call the method, Python will provide it.<br>
<br>
> * This particular variable refers to the object itself, and by convention, it is given the name self.

* Let’s define some attributes and methods to our ‘Vehicle’ Class:

In [125]:
class Vehicle:
    def __init__(self, numWheels, fuelType, seatCapacity, maxVelocity):
        self.numWheels = numWheels
        self.fuelType = fuelType
        self.seatCapacity = seatCapacity
        self.maxVelocity = maxVelocity

* As you can see we use the init method. We call it a constructor method. So when create the vehicle object, we can define this attributes.

* Let’s create an object and define attributes:

In [126]:
teslaS = Vehicle(4, 'electric', 5, 250)
teslaS

<__main__.Vehicle instance at 0x103e44560>

* Let’s create another car with methods this time

In [127]:
class Vehicle:
    def __init__(self, numWheels, fuelType, seatCapacity, maxVelocity):
        self.numWheels = numWheels
        self.fuelType = fuelType
        self.seatCapacity = seatCapacity
        self.maxVelocity = maxVelocity
    def speedUp(self, amount):
        self.maxVelocity += amount
    def printInfo(self):
        print("I have " + str(self.numWheels) + " wheels and can speed up to " + str(self.maxVelocity) + " km/h.")

    
myCar = Vehicle(4,'electric', 5, 120)

myCar.printInfo()

myCar.speedUp(100)

myCar.printInfo()

I have 4 wheels and can speed up to 120 km/h.
I have 4 wheels and can speed up to 220 km/h.
